MNT Refactor `_average_weighted_percentile` to avoid double sort

lucyleeow opened the pull request:
#### Reference Issues/PRs
Supercedes #30945


#### What does this implement/fix? Explain your changes.
Refactor `_average_weighted_percentile` so we are not just performing `_weighted_percentile` twice, thus avoids sorting and computing cumulative sum twice.

#30945 essentially uses the sorted indicies and calculates `_weighted_percentile(-array, 100-percentile_rank)` - this was verbose and required computing cumulative sum again on the negative (you could have used symmetry to avoid computing cumulative sum in cases when fraction above is greater than 0 - i.e., `g>0` from Hyndman and Fan)

I've followed the Hyndman and Fan computation more closely and calculate `g` and just use `j+1` (since we already know `j`). This did make handling the case where `j+1` had a sample weight of 0 (or when you have sample weight of 0 at the end of the array) more complex.

#### Any other comments?


<!--
Please be aware that we are a loose team of volunteers so patience is
necessary; assistance handling other issues is very welcome. We value
all user contributions, no matter how minor they are. If we are slow to
review, either the pull request needs some benchmarking, tinkering,
convincing, etc. or more likely the reviewers are simply busy. In either
case, we ask for your understanding during the review process.
For more information, see our FAQ on this topic:
https://scikit-learn.org/dev/faq.html#why-is-my-pull-request-not-getting-any-attention.

Thanks for contributing!
-->

lucyleeow committed:
try reverse cum sum

lucyleeow committed:
initial implementation, wip tests

lucyleeow committed:
fix and add tests, update use

lucyleeow commented on sklearn/utils/stats.py:
```diff
@@ -101,22 +114,55 @@ def _weighted_percentile(array, sample_weight, percentile_rank=50, xp=None):
+
+        result = xp.where(
+            is_fraction_above,
+            array[percentile_in_sorted, col_indices],
```
I initially thought this should be `percentile_plus_one_in_sorted` as from the paper, when g>0, $`\gamma=1`$, but `searchsorted` defaults to left (equals is on the right), whereas the paper defined $`j\le pn\lt j+1`$

ogrisel replied:
Maybe you could include the remark about the discrepancy between `searchsorted` and the paper convention as an inline comment in the code.

lucyleeow replied:
Done, hope it is adequate?

lucyleeow committed:
fixes and add tests

lucyleeow committed:
simplify zero sample code

lucyleeow committed:
typos

lucyleeow commented:
I think this is ready for review, maybe @ogrisel @betatim ?

lucyleeow commented:
I think this is ready for review, maybe @ogrisel @betatim ?

ogrisel committed:
Stronger tests

ogrisel reviewed (commented):
Thanks very much @lucyleeow for diving into this. I pushed a commit to make the randomized NumPy equivalence tests stronger and checked locally that they pass with all random seeds:

```
SKLEARN_TESTS_GLOBAL_RANDOM_SEED="all" pytest -vl sklearn/utils/tests/test_stats.py 
```

It seems that this PR fixes a rare edge case bug found in another PR (in addition to the CPU speed-up and memory improvements): https://github.com/scikit-learn/scikit-learn/pull/29641. We could write a changelog entry to document this. However, doing so would require crafting a minimal reproducer to precisely characterize the conditions under which this edge case can be triggered. Maybe we could have a generic changelog entry such as "improve CPU and memory usage in estimators and metric functions that rely on weighted percentiles and better handle edge cases" or something similar.

Please fix the conflicts and feel free to ping me again for the final review.

lucyleeow commented:
I looked into the failure from https://github.com/scikit-learn/scikit-learn/pull/29641#discussion_r2180443843 (I decided it was more relevant to detail here, and better that it does not get lost in that much bigger PR).

tl;dr the difference is due to floating point precision error, and the new implementation more closely matches how quantiles are implemented in numpy.

Using the test from https://github.com/scikit-learn/scikit-learn/pull/29641#discussion_r2180443843, I made a 'minimal' (enough) reproducer. Note that the percentile indices [9, 26, 36] differ between the 2 methods, but the example below just shows `9`.

```python

def test_weighted_percentiles():
    rng = np.random.RandomState(63)

    n_samples = 200
    X = rng.randn(n_samples, 3)
    sw = rng.randint(0, 5, size=n_samples)

    # repeat and compute using numpy non-weighted
    X_repeated = np.repeat(X, sw, axis=0)
    col_data = X_repeated[:, 0]
    sort_idx = np.argsort(col_data)
    col_data = col_data[sort_idx]
    percentiles = np.linspace(0, 100, num=49 + 1)
    percentiles = percentiles[1:-1]

    np_percentile = np.percentile( percentiles[9], method="averaged_inverted_cdf")
    print(f'{np_percentile=}')

    # Do not repeat and compute using our weighted_percentiles with weights
    col_data = X[:, 0]
    nnz_sw = sw != 0
    col_data = col_data[nnz_sw]
    sw = sw[nnz_sw]
    sort_idx = np.argsort(col_data)
    col_data = col_data[sort_idx]
    sw = sw[sort_idx]

    old = _averaged_weighted_percentile(col_data, sw, percentiles[9])
    print(f'{old=}')

    new = _weighted_percentile(col_data, sw, percentiles[9], average=True)
    print(f'{new=}')
```
In this case ([9]), the 'new' method (this PR): adjusted percentile rank (i.e. quantile * cumulative sum) ends up 80.0 exactly (I checked and it matches to 15 decimal places). This matches a value in the weighted_cdf (note sample weights are all integers), so fraction above is 0, and we take the average of the 2 adjacent values.

With the old method, when we take `_weighted_percentile` of `100-percentile`, adjusted percentile ends up being 312.00000000000006. Technically it should be 312 (i.e. cumulative sum - adjusted percentile rank of `percentile` (80)). This extra bit *just* pushes us to the next index, so `_weighted_percentile(percentile)` and `_weighted_percentile(100-percentile)` end up being the same number - but 100-percentile 'should' have been the adjacent value.

The other indices, [26, 36], have a similar problem where the adjusted percentile of `percentile` and `100-percentile` don't add up exactly to the cumulative sum (`weighted_cdf[-1]`), due to floating point precision. I guess the extra calculations (`100-percentile` then the `adjusted_percentile_rank = percentile_rank / 100 * weight_cdf[..., -1]`) gives more chance for error.

It's tricky to compare to numpy because they don't have a weighted implementation, and without weights you can just use `n` to calculate - which is what they do ([ref](https://github.com/numpy/numpy/blob/cbf648e398f0dcc1a1634e6689bc4de2e2cb9ccc/numpy/lib/_function_base_impl.py#L92)):

```python
_QuantileMethods = {
    # --- HYNDMAN and FAN METHODS
    # Discrete methods
    'inverted_cdf': {
        'get_virtual_index': lambda n, quantiles: _inverted_cdf(n, quantiles),  # noqa: PLW0108
        'fix_gamma': None,  # should never be called
    },
    'averaged_inverted_cdf': {
        'get_virtual_index': lambda n, quantiles: (n * quantiles) - 1,
        'fix_gamma': lambda gamma, _: _get_gamma_mask(
            shape=gamma.shape,
            default_value=1.,
            conditioned_value=0.5,
            where=gamma == 0),
    },
```

But ultimately I think this calculation `(n * quantiles) - 1` matches the new implementation more than reversing the array and calculating the 100-percentile.

I'll fix the merge conflict and address review now, thanks for review!

lucyleeow committed:
merge main

lucyleeow commented on sklearn/utils/stats.py:
```diff
@@ -7,8 +7,13 @@
+):
+    """Compute the weighted percentile.
+
+    Uses 'inverted_cdf' method when `average=False` (default) and
+    'averaged_inverted_cdf' when `average=True`.
```
Thanks, this is nicer

lucyleeow committed:
review

lucyleeow committed:
add whats new

lucyleeow commented on sklearn/utils/stats.py:
```diff
@@ -7,11 +7,14 @@
+):
+    """Compute the weighted percentile.
 
-    When the percentile lies between two data points of `array`, the function returns
-    the lower value.
```
I thought about replacing this sentence with a description of how `inverted_cdf` and `averaged_inverted_cdf` behaves but I sort of think the descriptions in the paper with the equations are clearer.

I also think it's initially 'confusing' when I describe that when a percentile falls exactly on a value in the array, we take the average of it and the next value in the array. It makes sense when you think about an array of 4 values, 50 percentile would fall exactly on the 2nd value, but of course symmetrized it makes sense to take the average of the 2nd and 3rd values.

I am happy to add this still if you think it is appropriate.

ogrisel replied:
Yes, please feel free to go ahead.

lucyleeow reviewed (commented):
Forgot to mention but I did an overhaul of the tests.

I removed one test because it was duplicated (checking that percentile 50 gives the same result as median) and one test that I added a while back that just did not make sense and was redundant.

Also added a general whats new entry. I think this is ready for review again @ogrisel

lucyleeow committed:
Merge branch 'main' into refactor_weighted_percentile

lucyleeow committed:
add comments

ogrisel commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,151 @@
+def test_weighted_percentile_ignores_zero_weight(
+    average, percentile_rank, expected_value
+):
     """Check `weighted_percentile(percentile_rank=0)` behaves correctly.
```
```suggestion
    """Edge case: 0 weight for the lowest and highest values and extreme ranks.
```

ogrisel commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,151 @@
-    # data
+@pytest.mark.parametrize("average", [True, False])
+@pytest.mark.parametrize("percentile_rank", [20, 35, 61])
+def test_weighted_median_frequency_weights(
```
```suggestion
def test_weighted_percentile_frequency_weight_semantics(
```

ogrisel commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,151 @@
+    percentile_repeated = _weighted_percentile(
+        x_repeated, np.ones_like(x_repeated), percentile_rank, average=average
+    )
+    assert percentile_weights == approx(percentile_repeated)
```
Please (re-)include the `percentile_rank=50` case and checks that it also matches the results of `np.median(x_repeated)` when `average is True`.

ogrisel replied:
```suggestion
    assert percentile_weights == approx(percentile_repeated)

    if percentile_rank == 50 and average:
        assert percentile_weights == approx(np.median(x_repeated))
```

lucyleeow replied:
Done

ogrisel reviewed (approved):
Thanks, @lucyleeow. Here is my (hopefully final) pass of review (mostly documentation fixes and a test improvement suggestion).

ogrisel commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -306,19 +349,34 @@ def test_weighted_percentile_all_nan_column():
+):
+    """Check `_weighted_percentile` equivalent to `np.quantile` with weights."""
+    # TODO: remove the following skip once no longer applicable.
+    if average and not uniform_weight:
+        pytest.skip(
+            "np.quantile does not support weights with method='averaged_inverted_cdf'"
```
The existing skip reason text sounds good enough to me.

lucyleeow commented on sklearn/utils/stats.py:
```diff
@@ -7,11 +7,21 @@
+    method when `average=False` (default) and 'averaged_inverted_cdf' when
+    `average=True`.
+
+    When the percentile lies exactly on a data point:
+    * 'inverted_cdf' takes the exact data point
+    * 'averaged_inverted_cdf' takes the average of the exact data point and the one
+      above it (this means it gives the same result as `median` for unit weights)
+    When the percentile lies between two data points, both methods take the higher
+    data point.
```
Hmm I think we are talking about different things. I am equating:

g = 0 : falls on exact value
g > 0 : falls between two data points

from the paper $`g=pn-j`$ (i.e. 'fraction above')
E.g., for array = 1, 2, 3, 4 and unit weights, the cdf would also be [1, 2, 3, 4]
pn = 0.5 * 4 = 2
Since `j` is 1-indexed, $`pn-j`$ is equal to 0, which matches:

```math
\gamma = \left\{
\begin{array}{ll}
\frac{1}{2} & \text{if } g = 0 \\
1 & \text{if } g > 0
\end{array}
\right.
```

as we do want to take the average between 2 data points when g=0.
It kind of also makes sense when you think of our old way of calculating it, when you reverse the array the 50 percentile would fall on '3', and they you take the average. 

BUT I can totally see the other interpretation, where 50 percentile is exactly half way between 2 and 3. But then how would you describe the case when the percentile is just above a data point? (e.g., percentile rank '55' in the same 4 data point array above)? That would also be 'between' two data points, but I guess not exactly half way, like the above case.

This is why I was not 100% keen on including this because it is confusing and NOT intuitive (and confused me a lot while I was working on this :sweat_smile: ) The paper with equations and figures is maybe nicer...?

lucyleeow commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,151 @@
-    score = _averaged_weighted_percentile(y, sw, 50)
+    With unit `sample_weight`, `_weighted_percentile` should match median except
+    when `average=False` and the number of samples is odd.
+    When number of samples is odd, `_weighted_percentile(average=False)` always falls
+    on a single observation (not between 2 values, in which case the lower value would
+    be taken) and is thus equal to `np.median`.
+    For an even number of samples, `median` gives the average between the 2 middle
+    samples, `_weighted_percentile(average=False)` gives the higher (right) sample.
```
I mucked up even and odd here :facepalm: but the same misunderstanding has been made as below https://github.com/scikit-learn/scikit-learn/pull/31775#discussion_r2323919910

<img width="1082" height="314" alt="image" src="https://github.com/user-attachments/assets/32487349-a83a-4605-b122-45d2f866ae67" />

The first method (inverted_cdf) only matches median when n is odd. It does not match when n is even.

lucyleeow commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,151 @@
-    y.fill(1.0)
-    sw = np.ones(102, dtype=np.float64)
-    sw.fill(0.0)
+def test_weighted_percentile_all_zero_weights():
```
Yeah I have an issue about this: https://github.com/scikit-learn/scikit-learn/issues/31032
please feel free to weigh in, I am happy to work on it.

ogrisel replied:
I edited the suggestion to link to the open issue.

lucyleeow committed:
review

lucyleeow commented:
Doc failure seems to be due to connection timeout?

ogrisel commented:
+1 for a long explanation with examples.

lucyleeow committed:
add explanation

lucyleeow committed:
update explanation

lucyleeow reviewed (commented):
Okay added more explanation and example. Let me know what you think @ogrisel !

ogrisel commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,151 @@
+    score = _weighted_percentile(y, sample_weight, 50, average=average)
 
+    # `_weighted_percentile(average=False)` does not match `median` when n is even
+    if size == 10 and average is False:
```
This makes the code both more explicit and still correct if we ever change the parametrization of `size`. 

```suggestion
    if size % 2 == 0 and average is False:
```

ogrisel commented on doc/whats_new/upcoming_changes/many-modules/31775.efficiency.rst:
```diff
@@ -0,0 +1,4 @@
+- Improved CPU and memory usage in estimators and metric functions that rely on
+  weighted percentiles and better match numpy and Scipy (un-weighted) implementations
```
```suggestion
  weighted percentiles and better match NumPy and SciPy (un-weighted) implementations
```

ogrisel reviewed (approved):
Thanks for the last iteration. This makes much more sense now, and it does not require the reader to be familiar with the details of Hyndman and Fan to understand the difference between the two methods.

A last pass of nitpicks / improvement suggestions.

lucyleeow commented:
@ogrisel while working on something related, I read: https://github.com/numpy/numpy/issues/8935#issuecomment-1382859616

I think the following should probably be true for our implementation of weights:

> Multiplying the weights by a constant does not change the quantile, i.e. w and c ∗ w give the same quantile value for c > 0 .

What do you think about adding a test for this?

lucyleeow committed:
review

lucyleeow committed:
add link

virchan commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,152 @@
-    sw.fill(0.0)
+# XXX: is this really what we want? Shouldn't we raise instead?
+# https://github.com/scikit-learn/scikit-learn/issues/31032
+def test_weighted_percentile_all_zero_weights():
```
Should we update this? It looks to me that returning `NaN` here was already decided in #31032.

virchan reviewed (commented):
LGTM! Thanks @lucyleeow! Just one naive question.

lucyleeow committed:
add constant multiplier test

virchan reviewed (approved):
CI is green, and the PR looks good to me. I'm also fine with having a third review before we merge.

ogrisel committed:
Trigger CI [all random seeds] test_weighted_percentile_constant_multiplier

lucyleeow committed:
use int constant

lucyleeow committed:
Trigger CI [all random seeds] test_weighted_percentile_constant_multi

lucyleeow commented:
I've re-triggered the all random seeds

OmarManzoor reviewed (commented):
LGTM aside from one point that I mentioned

lucyleeow committed:
fix test

OmarManzoor commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,175 @@
+def test_weighted_percentile_constant_multiplier(
+    global_random_seed, percentile_rank, average, constant
+):
+    """Check multiplying weights by constant does not change result.
```
```suggestion
    """Check multiplying weights by a constant does not change the result.
```

OmarManzoor commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,175 @@
-def test_weighted_median_integer_weights(global_random_seed):
-    # Checks average weighted percentile_rank=0.5 is same as median when manually weight
-    # data
+    Note scale invariance does not always hold when multiplying by a
```
```suggestion
    Note: scale invariance does not always hold when multiplying by a
```

OmarManzoor commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,175 @@
-def test_weighted_percentile_2d(global_random_seed):
+@pytest.mark.parametrize("average", [True, False])
+def test_weighted_percentile_2d(global_random_seed, average):
+    """Check `_weighted_percentile` behaviour correct when `array` is 2D."""
```
```suggestion
    """Check `_weighted_percentile` behaviour is correct when `array` is 2D."""
```

OmarManzoor commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -306,19 +373,34 @@ def test_weighted_percentile_all_nan_column():
+def test_weighted_percentile_like_numpy_quantile(
+    percentile, average, uniform_weight, global_random_seed
+):
+    """Check `_weighted_percentile` equivalent to `np.quantile` with weights."""
```
```suggestion
    """Check `_weighted_percentile` is equivalent to `np.quantile` with weights."""
```

OmarManzoor commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -329,24 +411,40 @@ def test_weighted_percentile_like_numpy_quantile(percentile, global_random_seed)
+def test_weighted_percentile_like_numpy_nanquantile(
+    percentile, average, uniform_weight, global_random_seed
+):
+    """Check `_weighted_percentile` equivalent to `np.nanquantile` with weights."""
```
```suggestion
    """Check `_weighted_percentile` is equivalent to `np.nanquantile` with weights."""
```

OmarManzoor reviewed (approved):
Thank you @lucyleeow 

I suggested some improvements in the docstrings and comments.

lucyleeow commented on sklearn/utils/tests/test_stats.py:
```diff
@@ -12,121 +12,175 @@
+# XXX: is this really what we want? Shouldn't we raise instead?
+# https://github.com/scikit-learn/scikit-learn/issues/31032
+def test_weighted_percentile_all_zero_weights():
+    """Check `weighted_percentile` with all weights equal to 0 returns last index."""
```
This would go over the line character limit, and it makes sense as is so I'm going to leave it.

lucyleeow committed:
review

lucyleeow commented:
@ogrisel are you happy for this to go in?

lucyleeow commented:
@ogrisel gentle ping

ogrisel commented:
Sorry for the slow feedback, thanks very much!
